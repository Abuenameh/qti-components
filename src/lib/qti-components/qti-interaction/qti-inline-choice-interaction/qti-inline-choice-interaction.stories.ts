import { html } from 'lit';
import { action } from '@storybook/addon-actions';

import './qti-inline-choice-interaction';

import { QtiInlineChoiceInteraction } from './qti-inline-choice-interaction';

export default {
  component: 'qti-inline-choice-interaction',
  argTypes: {
    inputWidthClass: {
      control: { type: 'select', options: QtiInlineChoiceInteraction.inputWidthClass },
      description: 'attr: class qti-inline-choice-interaction'
    },
    response: { description: 'prop: qti-assessment-item', type: 'string' },
    shuffle: { description: 'attr: qti-inline-choice-interaction', type: 'boolean' },
    readonly: { description: 'attr: qti-inline-choice-interaction', type: 'boolean' },
    disabled: { description: 'attr: qti-inline-choice-interaction', type: 'boolean' }
  }
};

export const Interaction = {
  render: args =>
    html` <qti-inline-choice-interaction
      @on-interaction-response="${e => {
        action(JSON.stringify(e.detail))();
      }}"
      .response=${args.response}
      ?disabled=${args.disabled}
      ?readonly=${args.readonly}
      class="${args.inputWidthClass}"
      response-identifier="RESPONSE"
      shuffle="${args.shuffle}"
    >
      <qti-inline-choice identifier="G">Gloucester</qti-inline-choice>
      <qti-inline-choice identifier="L">Lancaster</qti-inline-choice>
      <qti-inline-choice identifier="Y">York</qti-inline-choice>
    </qti-inline-choice-interaction>`
};

export const Item = {
  render: ({ shuffle, inputWidthClass, response }) =>
    html` <qti-assessment-item
      .responses="${response}"
      identifier="inline-choice"
      @qti-interaction-changed="${e => {
        action(JSON.stringify(e.detail))();
      }}"
    >
      <qti-item-body>
        <p>Identify the missing word in this famous quote from Shakespeare's Richard III.</p>
        <blockquote>
          <p>
            Now is the winter of our discontent <br />Made glorious summer by this sun of
            <qti-inline-choice-interaction
              class="${inputWidthClass}"
              response-identifier="RESPONSE"
              shuffle="${shuffle}"
            >
              <qti-inline-choice identifier="G">Gloucester</qti-inline-choice>
              <qti-inline-choice identifier="L">Lancaster</qti-inline-choice>
              <qti-inline-choice identifier="Y">York</qti-inline-choice>
            </qti-inline-choice-interaction>
            And all the clouds that lour'd upon our house <br />In the deep bosom of the ocean buried.
          </p>
        </blockquote>
      </qti-item-body>
    </qti-assessment-item>`
};
