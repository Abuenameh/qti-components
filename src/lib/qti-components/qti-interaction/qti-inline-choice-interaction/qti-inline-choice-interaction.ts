import { css, html } from 'lit';
import { property, state } from 'lit/decorators.js';
import { unsafeHTML } from 'lit/directives/unsafe-html.js';
import { Interaction } from '../internal/interaction/interaction';

interface OptionType {
  textContent: string;
  value: string;
  selected: boolean;
}
export class QtiInlineChoiceInteraction extends Interaction {
  public static inputWidthClass = [
    '',
    'qti-input-width-2',
    'qti-input-width-1',
    'qti-input-width-3',
    'qti-input-width-4',
    'qti-input-width-6',
    'qti-input-width-10',
    'qti-input-width-15',
    'qti-input-width-20',
    'qti-input-width-72'
  ];

  @state() options: OptionType[] = [];

  @state() correctOption: string = '';

  @property({ attribute: 'data-prompt', type: String })
  dataPrompt: string = 'select';

  static override get styles() {
    return [
      css`
        :host {
          display: inline-block;
        }
      `
    ];
  }

  override render() {
    return html`
      <select part="select" @change="${this.choiceSelected}" ?disabled="${this.disabled}" ?readonly="${this.readonly}">
        ${this.options.map(
          option => html`
            <option value="${option.value}" ?selected="${option.selected}">${unsafeHTML(option.textContent)}</option>
          `
        )}
      </select>

      ${unsafeHTML(this.correctOption)}
    `;
  }

  connectedCallback() {
    super.connectedCallback();
    this.addEventListener('on-dropdown-selected', this.choiceSelected);
    const choices = Array.from(this.querySelectorAll('qti-inline-choice'));
    this.options = [
      {
        textContent: this.dataPrompt,
        value: '',
        selected: false
      },
      ...choices.map(choice => ({
        textContent: choice.innerHTML,
        value: choice.getAttribute('identifier'),
        selected: false
      }))
    ];
  }

  disconnectedCallback() {
    this.removeEventListener('on-dropdown-selected', this.choiceSelected);
  }

  public validate(): boolean {
    const selectedOption = this.options.find(option => option.selected);
    return selectedOption ? selectedOption.value !== '' : false;
  }

  public reset() {
    this.options = this.options.map((option, i) => ({ ...option, selected: i === 0 }));
  }

  public set response(value: string) {
    this.options = this.options.map(option => {
      value === option.value && (option.selected = true);
      return option;
    });
  }

  set correctResponse(value: Readonly<string | string[]>) {
    if (value === '') {
      this.correctOption = '';
      return;
    }
    this.correctOption = `<span part="correct-option">${
      this.options.find(option => value === option.value).textContent
    }</span>`;
  }

  public choiceSelected(event: Event) {
    const selectedOptionValue = (event.target as HTMLSelectElement).value;
    this.options = this.options.map(option => ({ ...option, selected: option.value === selectedOptionValue }));
    this.saveResponse(selectedOptionValue);
  }
}

customElements.define('qti-inline-choice-interaction', QtiInlineChoiceInteraction);
