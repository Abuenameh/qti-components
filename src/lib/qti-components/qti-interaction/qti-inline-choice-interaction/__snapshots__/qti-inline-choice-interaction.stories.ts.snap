// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/qti-components/qti-interaction/qti-inline-choice-interaction DataPrompt smoke-test 1`] = `
<div id="root-inner">
  <qti-inline-choice-interaction response-identifier="RESPONSE"
                                 class
                                 shuffle
                                 data-prompt="Select the correct answer"
  >
    <qti-inline-choice identifier="G">
      Gloucester
    </qti-inline-choice>
    <qti-inline-choice identifier="L">
      Lancaster
    </qti-inline-choice>
    <qti-inline-choice identifier="Y">
      York
    </qti-inline-choice>
  </qti-inline-choice-interaction>
</div>
`;

exports[`lib/qti-components/qti-interaction/qti-inline-choice-interaction Default smoke-test 1`] = `
<div id="root-inner">
  <qti-inline-choice-interaction response-identifier="RESPONSE"
                                 class
                                 shuffle
  >
    <qti-inline-choice identifier="G">
      Gloucester
    </qti-inline-choice>
    <qti-inline-choice identifier="L">
      Lancaster
    </qti-inline-choice>
    <qti-inline-choice identifier="Y">
      York
    </qti-inline-choice>
  </qti-inline-choice-interaction>
</div>
`;

exports[`lib/qti-components/qti-interaction/qti-inline-choice-interaction Item smoke-test 1`] = `
<div id="root-inner">
  <qti-assessment-item identifier="inline-choice">
    <qti-item-body>
      <p>
        Identify the missing word in this famous quote from Shakespeare's Richard III.
      </p>
      <blockquote>
        <p>
          Now is the winter of our discontent
          <br>
          Made glorious summer by this sun of
          <qti-inline-choice-interaction response-identifier="RESPONSE"
                                         class
                                         shuffle
          >
            <qti-inline-choice identifier="G">
              Gloucester
            </qti-inline-choice>
            <qti-inline-choice identifier="L">
              Lancaster
            </qti-inline-choice>
            <qti-inline-choice identifier="Y">
              York
            </qti-inline-choice>
          </qti-inline-choice-interaction>
          And all the clouds that lour'd upon our house
          <br>
          In the deep bosom of the ocean buried.
        </p>
      </blockquote>
    </qti-item-body>
  </qti-assessment-item>
</div>
`;
